src/com/lid/intellij/translateme/actions/TranslateAction.java

```diff
@@ -12,114 +12,130 @@
-                    if (maxColumn < vis.column) {
-                        maxColumn = vis.column;
-                    }
-                }
-            }
-
-            VisualPosition pos = new VisualPosition(start.line, (start.column + maxColumn) / 2, false);
-
-            Point p = hintManager.getHintPosition(hint, editor, pos, HintManager.ABOVE);
-            hintManager.showEditorHint(hint, editor, p, flags, timeout, false);
-        }
-    }
+	static final NotificationGroup NOTIFICATION_GROUP = new NotificationGroup("TranslateMe Alerts", NotificationDisplayType.STICKY_BALLOON, false, null);
+
+	public TranslateAction() {
+		super(new TranslateHandler(new PopupActionHandler()));
+	}
+
+	public static String[] getLangPair() {
+		ConfigurationState state = PersistingService.getInstance().getState();
+
+		String from = state.getLangFrom();
+		String to = state.getLangTo();
+		return new String[] { from, to };
+	}
+
+	public static boolean isAutoDetect() {
+		ConfigurationState state = PersistingService.getInstance().getState();
+		return state.isAutoDetect();
+	}
+
+	public static boolean isTranslationTooltip() {
+		ConfigurationState state = PersistingService.getInstance().getState();
+		return state.isTranslationTooltip();
+	}
+
+	private static class PopupActionHandler implements ActionHandler {
+		@Override
+		public void handleResult(Editor editor, List<String> translated) {
+			Application app = ApplicationManager.getApplication();
+			app.invokeLater(() -> {
+				if (isTranslationTooltip()) {
+					StringBuilder sb = new StringBuilder();
+					int iMax = translated.size();
+					String sep = "";
+					for (int i = 0; i < iMax; i++) {
+						sb.append(sep).append(translated.get(i));
+						sep = "\n";
+					}
+					sb.append("<br><p><a href='http://translate.yandex.com/'>Powered by Yandex.Translator</a>");
+					showTooltip(editor, sb.toString());
+				} else {
+					ResultDialog resultDialog = new ResultDialog("Translated", translated);
+					resultDialog.setVisible(true);
+				}
+			});
+		}
+
+		@Override
+		public void handleError(Editor editor) {
+			Application app = ApplicationManager.getApplication();
+			app.invokeLater(() -> {
+				showErrorBalloon(editor, "Failed to translate");
+			});
+		}
+
+		private void showErrorBalloon(Editor editor, String message) {
+			NOTIFICATION_GROUP.createNotification("TranslateMe Error", XmlStringUtil.wrapInHtml(message), NotificationType.ERROR, (notification, event) -> {
+			}).notify(null);
+		}
+
+		static int getVisibleAreaWidth(Editor editor) {
+			Insets insets = editor.getContentComponent().getInsets();
+			int width = Math.max(0, editor.getScrollingModel().getVisibleArea().width - insets.left - insets.right);
+			return width;
+		}
+
+		private void showTooltip(Editor editor, String message) {
+			//String text = "<table><tr><td>&nbsp;</td><td>$message</td><td>&nbsp;</td></tr></table>"
+			HintManagerImpl hintManager = (HintManagerImpl) HintManager.getInstance();
```

Comment:
Please move this code to a separate class

Reply:
@vlivanov, I only made these changes because I found the original implementation with a modal dialog which could not be dismissed by escape or enter and required a mouse click on the tiny close button to be a serious disruption of workflow. Also, translating paragraphs was displayed as a single long line in the dialog, not always fitting on my desktop screen.

It was a quick fix that solved the problem.

I would love to spend more time on this plugin but I am overloaded with work and maintenance of other projects. If you could refactor the code to your liking it would be greatly appreciated. Otherwise, it will have to wait until I have a break in my schedule.

Reply:
No problem, thank you for your awesome contribution. I will refactor the class. Can you please still update the plugin version in this pr for it to be merged?

Reply:
@vlivanov, I will do it now. Otherwise it will get buried in the todo list and I will forget about it.

Reply:
@vlivanov, BTW I was not being sarcastic when I said I would love to spend more time on this plugin. 

I can see more productivity features:
* support multi-caret selections for translation and replacement.
* auto-select text under caret if no selection is present, 
   * if in a comment select the comment, 
   * if in a string select the string, etc.

   It is a pain to have to first select the text in a comment or a string. It could be done automatically by the plugin.

It is an awesome plugin to have when working with multi-language source code and documentation.

I am adding markdown document translation to my Markdown Navigator plugin. I used Yandex.Translate for my Laravel Translation Manager package for years and love their service but was too lazy and too busy to write my own Java code for it. Using your plugin by selecting and translating parts of a document got old really fast and I decided that I have to add full document translation to my plugin.

Thank you for a great plugin. If you are maintaining markdown documents please let me know. I have a standing offer of free license for Markdown Navigator to all plugin developers.